### DISCUSSION TITLE
Call function once child is done with rendering

### DISCUSSION THREAD

**Discussion by kilianso** [alpinejs/alpine#1273] [1. Help]
Hi 👋 I'm using Alpine.js to build an app that relies heavily on asynchronous API calls. So I fetch data and use `x-for` to loop and output things. This is all pretty cool and simple until you have to fire something, once a nested element is done rendering.

Here is an example (using PUG as the template language):

```pug
div(x-data='alpineHelper()' x-init='alpineTalkToApi()')
     template(x-if='data')
        div
           template(x-for='(entry, index) in data' :key='index')
              select(x-on:change='firesomeFunction()')
                 template(x-for='project in entry')
                    option(:value='project.id' x-text='project.name')
```
So what I'm doing here is basically just rendering multiple `select` elements based on API data. Now if you look at the "x-on:change" listener, I need to trigger a function whenever the select value changes. This works fine. My question is: how do I call this function immediately when the corresponding select menu is done rendering? Like "really done" when all values and options have been set. When I try to do this at the root level `x-init` it fires way too early and even with `$nextTick` it fires before the select menu value is set.

I tried using `x-show='firesomeFunction()'`, but this way the function is called many times, even though there's only one item in the outer loops array.

Is there some kind of equivalent to `x-init`, but on a nested level? Or is there any way to call a function at a sublevel without user interaction? What would be the best approach to accomplish something like this with Alpine?

Great project! Thanks for any tips.

**Comment by tvdr** [comment]
Hi,

According to the [docs for the magic helpers ](https://github.com/alpine-collective/alpine-magic-helpers#warning-using-componentparent-in-x-init ),  a no-duration setTimeout can help you with that

Check to console.logs in this [pen](https://codepen.io/tvdr/pen/YzNdVXE)

**Reply by KevinBatdorf** [reply]
I think he wants to wait until after it renders an update actually, but just on init.

**Comment by KevinBatdorf** [comment]
Try adding `:key=""` that are more unique than the index. 

By the way, you can format that code a bit nicer if you use 3 backticks and the name of the highlighting (``​`html)

Can you put up a codepen?

**Reply by kilianso** [reply]
Hey Kevin. What's the purpose of having more unique keys in this case?
I just updated the code above for better readability and yes, I'll do a Codepen asap.

Some more context: I'm building a time tracking app where I fetch data from 3 different API endpoints. The first one is `/track` to get tracked projects and packages of the current user. E.g. `[{pro: 1, pack: 4}, {pro: 2, pack: 1}]` The second endpoint is `/projects` to get all available projects in the company and the third one is `/packages/[projectID]` to get all packages of each project. So I fetch all projects and render them in a select dropdown, then I loop over all the projects and get the corresponding packages. According to the user data of `/track` I set the `selected` option in all the dropdowns.

Now I need to get the text of the currently selected option in each dropdown immediately once it is set and then again once a user changes a selection. If I do something like `x-show='() => console.log("ready")'` on a select element, I see this log way more often than it should be based on the length of the array. Seems like it is firing every time an option is added. And this seems a performance hit if I think of a long table with many select elements. So I just want to fire that log one time per select dropdown when it's ready and whenever its value is changed.

Anyway, as I said, I'll provide a code example asap.

**Reply by KevinBatdorf** [reply]
Adding the keys will make it easier for Alpine to find the correct element on render. They should be primitive types as well (like a string). If you're changing the data at all this should be used. May not help the described issue though, which will be easier to debug that after seeing it.
